### PR TITLE
Improve postgresql support

### PR DESCRIPTION
**Pull request by SquareMike**
@pierre Please review and merge.  Thank you.

### CODE REVIEW

**Comment by SquareMike** [comment]
I didn't realize the tests weren't working on this.  Any idea why I am getting this from the tests:

```  1) Killbill::Cybersource::PaymentPlugin should be able to charge a Credit Card directly and calls should be idempotent
     Failure/Error: @plugin.start_plugin
     Psych::SyntaxError:
       (<unknown>): expected <block end>, but found Key while parsing a block collection at line 11 column 3
     # ./spec/cybersource/remote/integration_spec.rb:11:in `block in (root)'
```

I am a Ruby newby.

**Comment by pierre** [comment]
> Psych::SyntaxError

It sounds like your YAML configuration is invalid.

File: Gemfile.lock [hunk 1]
```diff
@@ -140,10 +141,11 @@ PLATFORMS
...
   jdbc-mariadb (~> 1.1)
+  jdbc-postgres (~> 9.4)
   jdbc-sqlite3 (~> 3.7)
   killbill-cybersource!
   rake (>= 10.0.0)
   rspec (~> 2.12.0)
 
 BUNDLED WITH
-   1.11.2
+   1.13.6
```
Comment (pierre) [inline Gemfile.lock]: Bundler version needs to be kept at 1.11.2 (as set by our official build environment).

File: Gemfile.lock [hunk 2]
```diff
@@ -65,20 +65,21 @@ GEM
     diff-lcs (1.1.3)
     equalizer (0.0.11)
     erubis (2.7.0)
-    ethon (0.9.1)
+    ethon (0.10.1)
```
Comment (pierre) [inline Gemfile.lock]: These dependencies changes should be fine, but in general, unless required, let's keep the versions locked as-is.

File: Gemfile.lock [hunk 3]
```diff
@@ -1,7 +1,7 @@
 PATH
   remote: .
   specs:
-    killbill-cybersource (5.0.1)
+    killbill-cybersource (5.0.2)
```
Comment (pierre) [inline Gemfile.lock]: No need to change the version yet. The version bump is done by the build environment, right before a release (tag).

File: Jarfile.lock [hunk 4]
```diff
@@ -1,66 +1,68 @@
 ---
 :runtime:
-- io.dropwizard.metrics:metrics-core:jar:3.1.1
```
Comment (pierre) [inline Jarfile.lock]: There shouldn't be any `Jarfile.lock` change for this PR (the `Jarfile` dictates Kill Bill dependencies versions for tests, which haven't changed).

File: VERSION [hunk 5]
```diff
@@ -1 +1 @@
-5.0.1
+5.0.2
```
Comment (pierre) [inline VERSION]: Same remark as above.

File: cybersource.yml [hunk 6]
```diff
@@ -8,13 +8,17 @@
 #    :username: <%= ENV['OD_USERNAME'] %>
 #    :password: <%= ENV['OD_PASSWORD'] %>
 #    :check_for_duplicates: true
+  :test: true
```
Comment (pierre) [inline cybersource.yml]: `:test` is set to true by default. Also, it is already set on line 3 (per merchant account)?

File: pom.xml [hunk 7]
```diff
@@ -25,7 +25,7 @@
     <groupId>org.kill-bill.billing.plugin.ruby</groupId>
     <artifactId>cybersource-plugin</artifactId>
     <packaging>pom</packaging>
-    <version>5.0.1</version>
+    <version>5.0.2</version>
```
Comment (pierre) [inline pom.xml]: Same remark as above.

File: db/ddl.sql [hunk 8]
```diff
@@ -1,92 +1,92 @@
-CREATE TABLE `cybersource_payment_methods` (
```
Comment (pierre) [inline db/ddl.sql]: For reference, word-diff:

```
CREATE TABLE [-`cybersource_payment_methods`-]{+cybersource_payment_methods+} (
  [-`id` int(11) NOT NULL AUTO_INCREMENT,-]
[-  `kb_payment_method_id`-]{+id serial unique,+}
{+  kb_payment_method_id+} varchar(255) DEFAULT NULL,
  [-`token`-]{+token+} varchar(255) DEFAULT NULL,
  [-`cc_first_name`-]{+cc_first_name+} varchar(255) DEFAULT NULL,
  [-`cc_last_name`-]{+cc_last_name+} varchar(255) DEFAULT NULL,
  [-`cc_type`-]{+cc_type+} varchar(255) DEFAULT NULL,
  [-`cc_exp_month`-]{+cc_exp_month+} varchar(255) DEFAULT NULL,
  [-`cc_exp_year`-]{+cc_exp_year+} varchar(255) DEFAULT NULL,
  [-`cc_number`-]{+cc_number+} varchar(255) DEFAULT NULL,
  [-`cc_last_4`-]{+cc_last_4+} varchar(255) DEFAULT NULL,
  [-`cc_start_month`-]{+cc_start_month+} varchar(255) DEFAULT NULL,
  [-`cc_start_year`-]{+cc_start_year+} varchar(255) DEFAULT NULL,
  [-`cc_issue_number`-]{+cc_issue_number+} varchar(255) DEFAULT NULL,
  [-`cc_verification_value`-]{+cc_verification_value+} varchar(255) DEFAULT NULL,
  [-`cc_track_data`-]{+cc_track_data+} varchar(255) DEFAULT NULL,
  [-`address1`-]{+addresss+} varchar(255) DEFAULT NULL,
  [-`address2`-]{+address2+} varchar(255) DEFAULT NULL,
  [-`city`-]{+city+} varchar(255) DEFAULT NULL,
  [-`state`-]{+state+} varchar(255) DEFAULT NULL,
  [-`zip`-]{+zip+} varchar(255) DEFAULT NULL,
  [-`country`-]{+country+} varchar(255) DEFAULT NULL,
  [-`is_deleted` tinyint(1)-]{+is_deleted boolean+} NOT NULL DEFAULT '0',
  [-`created_at`-]{+created_at+} datetime NOT NULL,
  [-`updated_at`-]{+updated_at+} datetime NOT NULL,
  [-`kb_account_id`-]{+kb_account_id+} varchar(255) DEFAULT NULL,
  [-`kb_tenant_id`-]{+kb_tenant_id+} varchar(255) DEFAULT NULL,
  PRIMARY KEY [-(`id`),-]
[-  KEY `index_cybersource_payment_methods_on_kb_account_id` (`kb_account_id`),-]
[-  KEY `index_cybersource_payment_methods_on_kb_payment_method_id` (`kb_payment_method_id`)-]{+(id)+}
) {+/*!+} ENGINE=InnoDB CHARACTER SET utf8 COLLATE [-utf8_bin;-]{+utf8_bin */;+}
{+CREATE INDEX index_cybersource_payment_methods_kb_account_id ON cybersource_payment_methods(kb_account_id);+}
{+CREATE INDEX index_cybersource_payment_methods_kb_payment_method_id ON cybersource_payment_methods(kb_payment_method_id);+}
  
CREATE TABLE [-`cybersource_transactions`-]{+cybersource_transactions+} (
  [-`id` int(11) NOT NULL AUTO_INCREMENT,-]
[-  `cybersource_response_id` int(11)-]{+id serial unique,+}
{+  cybersource_response_id bigint /*! unsigned */+} NOT NULL,
  [-`api_call`-]{+api_call+} varchar(255) NOT NULL,
  [-`kb_payment_id`-]{+kb_payment_id+} varchar(255) NOT NULL,
  [-`kb_payment_transaction_id`-]{+kb_payment_transaction_id+} varchar(255) NOT NULL,
  [-`transaction_type`-]{+transaction_type+} varchar(255) NOT NULL,
  [-`payment_processor_account_id`-]{+payment_processor_account_id+} varchar(255) DEFAULT NULL,
  [-`txn_id`-]{+txn_id+} varchar(255) DEFAULT NULL,
  [-`amount_in_cents` int(11)-]{+amount_in_cents int+} DEFAULT NULL,
  [-`currency`-]{+currency+} varchar(255) DEFAULT NULL,
  [-`created_at`-]{+created_at+} datetime NOT NULL,
  [-`updated_at`-]{+updated_at+} datetime NOT NULL,
  [-`kb_account_id`-]{+kb_account_id+} varchar(255) NOT NULL,
  [-`kb_tenant_id`-]{+kb_tenant_id+} varchar(255) NOT NULL,
  PRIMARY KEY [-(`id`),-]
[-  KEY `index_cybersource_transactions_on_kb_payment_id` (`kb_payment_id`),-]
[-  KEY `index_cybersource_transactions_on_cybersource_response_id` (`cybersource_response_id`)-]{+(id)+}
) {+/*!+} ENGINE=InnoDB CHARACTER SET utf8 COLLATE [-utf8_bin;-]{+utf8_bin */;+}
{+CREATE INDEX index_cybersource_transactions_kb_payment_id ON cybersource_transactions(kb_payment_id);+}
{+CREATE INDEX index_cybersource_transactions_cybersource_response_id ON cybersource_transactions(cybersource_response_id);+}

CREATE TABLE [-`cybersource_responses`-]{+cybersource_responses+} (
  [-`id` int(11) NOT NULL AUTO_INCREMENT,-]
[-  `api_call`-]{+id serial unique,+}
{+  api_call+} varchar(255) NOT NULL,
  [-`kb_payment_id`-]{+kb_payment_id+} varchar(255) DEFAULT NULL,
  [-`kb_payment_transaction_id`-]{+kb_payment_transaction_id+} varchar(255) DEFAULT NULL,
  [-`transaction_type`-]{+transaction_type+} varchar(255) DEFAULT NULL,
  [-`payment_processor_account_id`-]{+payment_processor_account_id+} varchar(255) DEFAULT NULL,
  [-`message`-]{+message+} text DEFAULT NULL,
  [-`authorization`-]{+authorisation+} varchar(255) DEFAULT NULL,
  [-`fraud_review` tinyint(1)-]{+fraud_review boolean+} DEFAULT NULL,
  [-`test` tinyint(1)-]{+test boolean+} DEFAULT NULL,
  [-`params_merchant_reference_code`-]{+params_merchant_reference_code+} varchar(255) DEFAULT NULL,
  [-`params_request_id`-]{+params_request_id+} varchar(255) DEFAULT NULL,
  [-`params_decision`-]{+params_decision+} varchar(255) DEFAULT NULL,
  [-`params_reason_code`-]{+params_reason_code+} varchar(255) DEFAULT NULL,
  [-`params_request_token`-]{+params_request_token+} varchar(255) DEFAULT NULL,
  [-`params_currency`-]{+params_currency+} varchar(255) DEFAULT NULL,
  [-`params_amount`-]{+params_amount+} varchar(255) DEFAULT NULL,
  [-`params_authorization_code`-]{+params_authorization_code+} varchar(255) DEFAULT NULL,
  [-`params_avs_code`-]{+params_avs_code+} varchar(255) DEFAULT NULL,
  [-`params_avs_code_raw`-]{+params_avs_code_raw+} varchar(255) DEFAULT NULL,
  [-`params_cv_code`-]{+params_cv_code+} varchar(255) DEFAULT NULL,
  [-`params_authorized_date_time`-]{+params_authorized_date_time+} varchar(255) DEFAULT NULL,
  [-`params_processor_response`-]{+params_processor_response+} varchar(255) DEFAULT NULL,
  [-`params_reconciliation_id`-]{+params_reconciliation_id+} varchar(255) DEFAULT NULL,
  [-`params_subscription_id`-]{+params_subscription_id+} varchar(255) DEFAULT NULL,
  [-`avs_result_code`-]{+avs_result_code+} varchar(255) DEFAULT NULL,
  [-`avs_result_message`-]{+avs_result_message+} varchar(255) DEFAULT NULL,
  [-`avs_result_street_match`-]{+avs_result_street_match+} varchar(255) DEFAULT NULL,
  [-`avs_result_postal_match`-]{+avs_result_postal_match+} varchar(255) DEFAULT NULL,
  [-`cvv_result_code`-]{+cvv_result_code+} varchar(255) DEFAULT NULL,
  [-`cvv_result_message`-]{+cvv_result_message+} varchar(255) DEFAULT NULL,
  [-`success` tinyint(1)-]{+success boolean+} DEFAULT NULL,
  [-`created_at`-]{+created_at+} datetime NOT NULL,
  [-`updated_at`-]{+updated_at+} datetime NOT NULL,
  [-`kb_account_id`-]{+kb_account_id+} varchar(255) DEFAULT NULL,
  [-`kb_tenant_id`-]{+kb_tenant_id+} varchar(255) DEFAULT NULL,
  PRIMARY KEY [-(`id`),-]
[-  KEY `index_cybersource_responses_on_kb_payment_id_kb_tenant_id` (`kb_payment_id`, `kb_tenant_id`)-]{+(id)+}
) {+/*!+} ENGINE=InnoDB CHARACTER SET utf8 COLLATE [-utf8_bin;-]{+utf8_bin */;+}
{+CREATE INDEX index_cybersource_responses_kb_payment_id_kb_tenant_id ON cybersource_responses(kb_payment_id, kb_tenant_id);+}
```

File: Rakefile [hunk 9]
```diff
@@ -1,5 +1,12 @@
 #!/usr/bin/env rake
 
+module TempFixForRakeLastComment
```
Comment (pierre) [inline Rakefile]: Is that for the deprecation warning? If so, let's do this in a separate PR. It is also unclear if it belongs here or in the base framework (which would solve it for all plugins).

File: Gemfile.lock (hunk 2 above)
Comment (SquareMike) [inline Gemfile.lock]: These got bumped by the Ruby system, I am compiling with:
```bundle install; jbundle install
bundle exec rake killbill:clean;bundle exec rake build;bundle exec rake killbill:package```

Is there a better way that doesn't result in the versions changing?

File: Rakefile (hunk 9 above)
Comment (SquareMike) [inline Rakefile]: My local Ruby system must be higher than the one our build is using because I can't compile with last_comment in place.  Is there a Ruby command I should be using to compile using libraries this code expects?

File: Gemfile.lock (hunk 2 above)
Comment (pierre) [inline Gemfile.lock]: > Is there a better way that doesn't result in the versions changing?

`bundle install` should not update the Gemfile.lock (`bundle update` would though).

In any case, assuming you have all dependencies installed (after running `bundle install`), you can `git add -p` the relevant hunk for PostgreSQL and `git checkout` the file.

There is usually no issue running slightly different versions (i.e. it shouldn't impact your testing), but for the official version / build, we like to be prudent and keep a dependency tree that we have tested ourselves.

File: Rakefile (hunk 9 above)
Comment (pierre) [inline Rakefile]: > My local Ruby system must be higher

Which version are you using? We typically run JRuby 1.7.2x, see http://docs.killbill.io/latest/plugin_development.html#_ruby_plugins and https://github.com/killbill/killbill-cloud/blob/master/docker/templates/build/Dockerfile#L25 (official build image).

Comment (SquareMike) [inline Rakefile]: 2.3.1.  I got it figured out though, if I revert the changes to the lock files and use bundle properly then it will use a compiler that matches our dependencies.  I pushed a commit that backs all of that out.  Thanks for bearing with me while I figure out the build system.

**Comment by SquareMike** [comment]
K, I think all the changes from the first review are in.

**Comment by pierre** [comment]
:1st_place_medal:
